pupynere.py

```diff
@@ -1056,7 +1056,7 @@ def __setitem__(self, index, data):
                 recs = rec_index + 1
             if recs > len(self.data):
                 shape = (recs,) + self._shape[1:]
-                self.data.resize(shape)
+                self.data.resize(shape, refcheck=False)
```

Comment:
What was your reasoning behind changing this?

Reply:
There was an error when I ran the tests with `pytest --cov=. tests/`
```bash
E               ValueError: cannot resize an array that references or is referenced
E               by another array in this way.
E               Use the np.resize function or refcheck=False
```

Reply:
Hm. Can you point me to the test that was getting that error message?

Reply:
`TestGeneratorRecvars` in  `tests/test_generators.py` and `FirstUnlimdimTestCase` in `tests/test_unlim.py` but it only happens with `pytest --cov=. tests/` and not `pytest`
```bash
________________________________________ FirstUnlimdimTestCase.runTest ________________________________________

self = <tests.test_unlim.FirstUnlimdimTestCase testMethod=runTest>

    def setUp(self):
        self.file = tempfile.mktemp(".nc")
        f  = pupynere.netcdf_file(self.file, 'w')
        # foo has a single unlimited dimension
        f.createDimension('n1', None)
        f.createDimension('n2', n2dim)
        f.createDimension('n3', n3dim)
        foo = f.createVariable('data1', ranarr.dtype.str[1:], ('n1','n2','n3'))
        # write some data to it.
>       foo[:] = ranarr

tests/test_unlim.py:48: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <pupynere.netcdf_variable object at 0x7fe597ba03a0>, index = slice(None, None, None)
data = array([[[0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0...[0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0., 0., 0., 0.]]])

    def __setitem__(self, index, data):
        if not self._data_allocated():
            self._allocate_data()
    
        if self.maskandscale:
            missing_value = (
                    self._attributes.get('missing_value') or
                    self._attributes.get('_FillValue') or
                    getattr(data, 'fill_value', 999999))
            self._attributes.setdefault('missing_value', missing_value)
            self._attributes.setdefault('_FillValue', missing_value)
            data = np.ma.asarray(data).filled(missing_value)
            data = (data - self._attributes.get('add_offset', 0.0)) / self._attributes.get('scale_factor', 1.0)
    
        # Expand data for record vars?
        if self.isrec:
            if isinstance(index, tuple):
                rec_index = index[0]
            else:
                rec_index = index
            if isinstance(rec_index, slice):
                recs = (rec_index.start or 0) + len(data)
            else:
                recs = rec_index + 1
            if recs > len(self.data):
                shape = (recs,) + self._shape[1:]
>               self.data.resize(shape)
E               ValueError: cannot resize an array that references or is referenced
E               by another array in this way.
E               Use the np.resize function or refcheck=False

pupynere.py:1059: ValueError

```

Reply:
Hm, I don't like turning off reference counting checks - they're pretty useful and might catch a bug for us in the future, and any bug a reference count helps you catch would be an utter nightmare without them.

I *think* (or at least hope) the problem in this case isn't bad memory handling in pupynere-pdp, just a quirk of the way the tests in question are set up. The tests supply data to the netCDF variable by repeating the same block of data, `ranarr`, twice. Unfortunately, I have been having trouble getting the tests running on my machine and can't experiment with test setup as easily as I'd like.

Reply:
I removed the `refcheck=False` but can't reproduce the test failure on either my workstation or on a separate branch in github. This makes it rather difficult to figure out whether the error from numpy relates to the way the tests are structured (no big deal, we just rewrite them) or the way pupynere is structured (might bite us in the butt someday and it will be Very Bad).

Do you have any ideas for how I could reproduce the error? @jameshiebert ?

If not, I guess we should go ahead with the merge - we don't know of anything broken, just ... breakable. Please add a comment to `pupynere.py` so some hypothetical acciursed future programmer trying to track down a very subtle bug that intermittently scrambles the order of variable data blocks will have an easier time noticing that some of numpy's checks intended to prevent that sort of thing have been disabled. :)

Reply:
I couldn't reproduce the error either (I tried Python 3.7 and 3.8), so it's tempting to omit this change.

However, the better option would be to use the first choice from the error message that you saw and to use the module level `np.resize()` function instead of the object level `[array].resize()` function, that is being used now.

I haven't done a full audit of the code, but reading over [the docs](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.resize.html#numpy.ndarray.resize) I wasn't overly worried. That said, @corviday is definitely right to be concerned for subtle bugs, so this shouldn't be dismissed. Hopefully a solution would be to use the module level `np.resize()`; maybe it will have more internal checks that require us to give up on the reference checking?

Give it a shot and see how it plays.

Reply:
@jameshiebert I did try that first and it raises a different error, but I can poke at it more if it is worth it.
```bash
                np.resize(self.data, shape)
>       self.data[index] = data
E       ValueError: could not broadcast input array from shape (4,10,8) into shape (0,10,8)

```

Reply:
So this works 
```
        self.data = np.resize(self.data, shape)
        self.data[index] = data
```
but it changed the filesize in one of the tests
```
>               assert nc.filesize == 80116
E               AssertionError: assert 85252 == 80116
E                +  where 85252 = <pupynere.netcdf_file object at 0x7f2cd01c0bb0>.filesize
```